### ISSUE TITLE
How do strings work?

### ISSUE THREAD

**Issue by codereport** [codereport/jello#13]
@c4augustus @lyxal @UnrelatedString Seeing as via streams / issues I have found some people that know about Jelly, how do strings work? Supposedly the back double quote is for characters, and an enclose front/back double quote is for strings. But for the life of me it doesn't seem to work. To check if the character A exists in string ABC:

![image](https://github.com/codereport/jello/assets/36027403/12222c05-e7e2-4fb9-8401-60e1af1a481b)

but this works:

![image](https://github.com/codereport/jello/assets/36027403/04c42013-2def-4cea-b217-9c768abddd1c)

**Comment by c4augustus** [comment]
You are right that something is wrong with Jelly's parsing with what/how we pass the expression to it:
<img width="430" alt="Captura de ecrã 2023-12-16, às 17 28 08" src="https://github.com/codereport/jello/assets/5049680/13f746db-57a9-41ae-81ed-eda1cc64a064">

But Jelly Balls actually resolves this if we do NOT provide the weird “ ” glyphs. It seems to assume that any alpha character sequence is a string and shows it as a string wrapped with conventional " " glyphs, and it is able to pass something to jelly to get the correct answers.
<img width="357" alt="Captura de ecrã 2023-12-16, às 17 13 15" src="https://github.com/codereport/jello/assets/5049680/8d76adce-7c6b-4476-9a26-0e762edff429">
<img width="347" alt="Captura de ecrã 2023-12-16, às 17 14 04" src="https://github.com/codereport/jello/assets/5049680/461c1a97-b740-4571-a4f4-cb37c4f4be64">
<img width="345" alt="Captura de ecrã 2023-12-16, às 17 14 17" src="https://github.com/codereport/jello/assets/5049680/056eefcf-0666-463c-b4d8-681ad19f2a23">

I am investigating the Jelly Balls JS code to see how it is doing this. Didn't you mention that Adam wrote Jelly  Balls?

**Comment by c4augustus** [comment]
Actually, your example does work properly from the command line:
<img width="403" alt="Captura de ecrã 2023-12-16, às 17 40 42" src="https://github.com/codereport/jello/assets/5049680/78b3cee7-b3cd-4caa-8efa-ff6c944df112">

But Jelly Beans is doing it's own thing to decide what is a string when entered from the JS console, which it is then probably wrapping whatever you enter inside another set of weird “...” quote glyphs when it calls jelly. Notice how it is showing Args at the bottom of each string:
 
<img width="334" alt="Captura de ecrã 2023-12-16, às 17 44 12" src="https://github.com/codereport/jello/assets/5049680/74f7308b-0b5a-4763-9f5d-7c4b2b24b866">
<img width="355" alt="Captura de ecrã 2023-12-16, às 17 44 41" src="https://github.com/codereport/jello/assets/5049680/763698f7-5105-4824-b9b3-a44370239c7d">
<img width="351" alt="Captura de ecrã 2023-12-16, às 17 44 59" src="https://github.com/codereport/jello/assets/5049680/214c0602-4ea4-4bc6-a8b5-134830f5b083">
<img width="347" alt="Captura de ecrã 2023-12-16, às 17 45 38" src="https://github.com/codereport/jello/assets/5049680/f0edc78d-a1b7-4ceb-ae62-e91250e99520">
<img width="350" alt="Captura de ecrã 2023-12-16, às 17 46 30" src="https://github.com/codereport/jello/assets/5049680/7b19895d-e998-4ca3-b61e-62732899fb09">

**Comment by c4augustus** [comment]
Okay, I found the problem. It's the jelly program itself when called from a shell command-line, or through python subprocess.run as jello is doing. I noticed that in what I wrote above I was passing the arguments to the chain as part of the chain rather than input arguments as you are doing. When we do this, jelly/__main__.py splits up all of the characters of the input arguments into an array of individual characters that INCLUDE the quotes into a new array of isolated characters. It also puts a single character into an array of length one which fails to be found in an array of scalar characters (last example below). I am continuing to dig into this.
<img width="1164" alt="Captura de ecrã 2023-12-17, às 13 38 36" src="https://github.com/codereport/jello/assets/5049680/c8ff7d53-9ccd-40b4-ac03-32965bd0b632">

**Comment by c4augustus** [comment]
I figured out why it's not working and I have made a screen recording to explain it here: https://www.youtube.com/watch?v=rvUssFi9qQU

**Comment by codereport** [comment]
@c4augustus this is awesome! I am going to do a live stream focusing on strings and simplifying / fixing them in Jellyfish : )

**Comment by UnrelatedString** [comment]
Haven't watched the video (and sorry for the late reply), but the issue here is that Jelly doesn't parse arguments to the program the same way as it parses literals within the program: program arguments are evaluated as Python then recursively "jellified". `”ae“abc”` will return `1`, but since jellification treats all Python strings as strings (rather than characters), there's no way to pass a character and a string to the program `e` that returns `1`, because there's no way to pass a single character in at all!

I actually have no idea how Jelly Balls is ending up doing that anyways in some cases, but it wouldn't surprise me if it was designed to have some way to input characters so users don't have to work around it like usual.

**Comment by c4augustus** [comment]
> @c4augustus this is awesome! I am going to do a live stream focusing on strings and simplifying / fixing them in Jellyfish : )

I will attend if I can.

**Comment by UnrelatedString** [comment]
Oh, incidentally, if you ever want to drop by [Jelly Hypertraining](https://chat.stackexchange.com/rooms/57815/jelly-hypertraining), just let a mod know and they can unfreeze it.
